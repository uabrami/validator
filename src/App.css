
.App {
  align-items: center;
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
}

.App-main {
  min-height: 100vh;
}
.App-input {
  border-radius: 0.3rem;
  border: 0.1rem solid #d3d3d3;
  height: 2rem;
  margin-right: 0.5rem;
  width: 20rem;
}
p {
  font-size: 1rem;
  margin-bottom: 0.3rem;
}
button, input[type=submit] {
  height: 2.3rem;
  border-radius: 0.3rem;
  margin-right: 0.5rem;
}
button {
  background-color:#d3d3d3;
}
input[type=submit] {
  background-color: #F07100;
}
