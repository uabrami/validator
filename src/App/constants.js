export const CONFIRMATION_MESSAGES = {
  isNotValidMessage:
    "Your password does not meet the requirements, please try again.",
  isValidMessage: "Congrats, you have successfully submitted your password!",
};

export default CONFIRMATION_MESSAGES;
