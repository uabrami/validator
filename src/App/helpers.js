/**
 * hasCorrectLimit
 * @description checks the char limit of the password
 */
export const hasCorrectLimit = (str) => str.length >= 8 && str.length <= 72;

/**
 * hasLowerCase
 * @description checks if the password has a lowercase char
 */
export const hasLowerCase = (str) => {
  return str.toUpperCase() !== str;
};

/**
 * hasNumber
 * @description check if the password has a number
 */
export const hasNumber = (str) => {
  return /\d/.test(str);
};

/**
 * hasUpperCase
 * @description check if the password has an uppercase char
 */
export const hasUpperCase = (str) => {
  return str.toLowerCase() !== str;
};

/**
 * isCompleteMessage
 * @description check if a requirement has been met
 */
export const isCompleteMessage = (value, user, messages, setMessages) => {
  const newMessages = [...messages];
  newMessages[0].isCompleted = hasCorrectLimit(value) ? true : false;
  newMessages[1].isCompleted = hasLowerCase(value) ? true : false;
  newMessages[2].isCompleted =
    !usedEmailAddress(value, user) && value.length ? true : false;
  newMessages[3].isCompleted = hasUpperCase(value) ? true : false;
  newMessages[4].isCompleted = hasNumber(value) ? true : false;
  setMessages(newMessages);
};

/**
 * isPasswordValid
 * @description checks if the password meets all the requirements and is valid
 */

export const isPasswordValid = (value, user) => {
  return (
    hasCorrectLimit(value) &&
    hasLowerCase(value) &&
    !usedEmailAddress(value, user) &&
    hasUpperCase(value) &&
    hasNumber(value)
  );
};

/**
 * usedEmailAddress
 * @description check if the password does not include the email username
 */
export const usedEmailAddress = (str, user) => {
  const emailFragment = user.substring(
    user.indexOf("email: "),
    user.indexOf("@")
  );
  const username = emailFragment.substring(8);
  return str.includes(username);
};
