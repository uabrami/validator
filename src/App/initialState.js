export const initialStateMessages = [
  {
    text: '8-72 Characters',
    isCompleted: false,
  },
  {
    text: '1 Lowercase Character',
    isCompleted: false,
  },
  {
    text: 'Should Not Match Your Email Address',
    isCompleted: false,
  },
  {
    text: '1 Uppercase Character',
    isCompleted: false
  },
  {
    text: '1 Number',
    isCompleted: false
  },
];