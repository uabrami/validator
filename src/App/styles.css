
.App {
  align-items: center;
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
  margin-top: 5rem;
}

.checkbox {
  background-color: white;
  margin-right: 0.3rem;
  vertical-align: middle;
}

.checkboxTitle {
  display: inline-block;
  font-size: 0.8rem;
  vertical-align: middle;
}

.confirmationMessage {
  font-size: 0.8rem;
}

.fade-wrapper {
  transition:opacity ease-in-out;
}
.fade-out {
  opacity:0;
}
.fade-in {
  opacity:1;
}

.labelConfMessageDiv {
  align-items: center;
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  width: 100%;
  margin-bottom: 0.3rem;
}

.form {
  align-items: center;
  display: flex;
  flex-direction: row;
  justify-content: flex-start;
}

.passwordInput {
  border-radius: 0.3rem;
  border: 0.1rem solid #d3d3d3;
  height: 2rem;
  margin-right: 0.5rem;
  padding: 0 0.5rem;
  width: 30rem;
}

.showPasswordContainer {
  align-items: center;
  border-radius: 0.3rem;
  border: 0.1rem solid #d3d3d3;
  display: flex;
  flex-direction: row;
  height: 2rem;
  margin-right: 0.5rem;
  padding: 0 0.2rem;
  width: 3.6rem;
}

input[type=submit] {
  background-color: #F07100;
  border-radius: 0.3rem;
  border: 0.1rem solid #d3d3d3;
  height: 2.2rem;
}

input[type=submit]:hover {
  background-color: black;
  color: white;
}

input[type=submit]:focus, input[type="checkbox"]:focus, input:focus {
  outline: none;
}

p {
  margin: 0;
}
