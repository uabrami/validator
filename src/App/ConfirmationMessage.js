import React from "react";
import "./styles.css";

const ConfirmationMessage = ({ isValid, message }) =>{
  const isValidColor = isValid ? '#61B329' : '#DC143C';
 return (
    <p
      className='confirmationMessage'
      style={{color: isValidColor}}
    >
      {message}
    </p>
);
 }

export default ConfirmationMessage;