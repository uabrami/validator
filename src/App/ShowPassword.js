import React from "react";
import "./styles.css";

const ShowPassword = ({ handleToggle }) => (
  <div className="showPasswordContainer">
    <input
      className="checkbox"
      type="checkbox"
      onClick={handleToggle}
    />
    <span className="checkboxTitle">Show</span>
  </div>
);

export default ShowPassword;