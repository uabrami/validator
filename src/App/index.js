import React, { useEffect, useState } from "react";
import axios from "axios";

import ShowPassword from "./ShowPassword";
import ValidationMessages from "../ValidationMessages";
import ConfirmationMessage from "./ConfirmationMessage";
import { initialStateMessages } from "./initialState";
import { isPasswordValid, isCompleteMessage } from "./helpers";
import CONFIRMATION_MESSAGES from "./constants";
import "./styles.css";

const { isNotValidMessage, isValidMessage } = CONFIRMATION_MESSAGES;

const App = () => {
  const [confMessage, setConfMessage] = useState("");
  const [fadeState, setfadeState] = useState("fade-out");
  const [messages, setMessages] = useState(initialStateMessages);
  const [showText, setShowText] = useState(false);
  const [user, setUser] = useState("");
  const [value, setValue] = useState("");

  const fadeDuration = 1500;
  const isValid = isPasswordValid(value, user);
  const messageDuration = 2500;

  useEffect(() => {
    const fetchData = async () => {
      const config = {
        method: "post",
        url: "http://www.mocky.io/v2/5de6c328370000a21d0925f2",
        headers: { "Content-Type": "application/json; charset=UTF-8" },
      };
      const result = await axios(config);
      setUser(result.data);
    };
    fetchData();
  }, []);

  useEffect(() => {
    isCompleteMessage(value, user, messages, setMessages);
    // eslint-disable-next-line
  }, [value]);

  const handleChange = (e) => {
    setValue(e.target.value);
  };

  const handleSubmit = (e) => {
    e.preventDefault();
    if (isValid) {
      setConfMessage(isValidMessage);
      setfadeState("fade-in");
    } else {
      setConfMessage(isNotValidMessage);
      setfadeState("fade-in");
    }
    setTimeout(() => {
      setfadeState("fade-out");
    }, messageDuration);
  };

  const handleToggle = () => setShowText(!showText);
  return (
    <div className="app">
      <main>
        <div className="messageWrapper">
          <p>Password</p>
          <div
            className={`fade-wrapper ${fadeState}`}
            style={{ transitionDuration: `${fadeDuration}ms` }}
          >
            <ConfirmationMessage isValid={isValid} message={confMessage} />
          </div>
        </div>
        <form className="form" onSubmit={handleSubmit}>
          <input
            aria-label="input password"
            className="passwordInput"
            type={showText ? "text" : "password"}
            value={value}
            onChange={handleChange}
          />
          <ShowPassword
            aria-label="show password text"
            handleToggle={handleToggle}
          />
          <input aria-label="submit password" type="submit" value="Submit" />
        </form>
        <ValidationMessages
          aria-label="password requirements"
          messages={messages}
        />
      </main>
    </div>
  );
};

export default App;
