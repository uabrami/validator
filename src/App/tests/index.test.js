import React from "react";
import App from "../index";
import { render } from "@testing-library/react";

import { checkBasicRendering } from "../../testingHelpers";

const Component = <App />;

describe("<App />", () => {
  it("renders expected content", () => {
    const { container } = render(Component);
    expect(container.querySelectorAll("main").length).toEqual(1);
    expect(container.querySelectorAll("form").length).toEqual(1);
    expect(container.querySelectorAll("p").length).toEqual(2);
    expect(container.querySelectorAll("input").length).toEqual(3);
  });
  it("Should render and match the snapshot", () => {
    const {
      container: { firstChild },
    } = render(Component);
    expect(firstChild).toMatchSnapshot();
  });
  checkBasicRendering(Component);
});
