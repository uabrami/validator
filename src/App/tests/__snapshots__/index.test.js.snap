// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> Should render and match the snapshot 1`] = `
<div
  class="app"
>
  <main>
    <div
      class="messageWrapper"
    >
      <p>
        Password
      </p>
      <div
        class="fade-wrapper fade-out"
        style="transition-duration: 1500ms;"
      >
        <p
          class="confirmationMessage"
          style="color: rgb(220, 20, 60);"
        />
      </div>
    </div>
    <form
      class="form"
    >
      <input
        aria-label="input password"
        class="passwordInput"
        type="password"
        value=""
      />
      <div
        class="showPasswordContainer"
      >
        <input
          class="checkbox"
          type="checkbox"
        />
        <span
          class="checkboxTitle"
        >
          Show
        </span>
      </div>
      <input
        aria-label="submit password"
        type="submit"
        value="Submit"
      />
    </form>
    <div
      class="messagesContainer"
    >
      <ul>
        <li>
          <span>
            8-72 Characters
          </span>
        </li>
        <li>
          <span>
            1 Lowercase Character
          </span>
        </li>
        <li>
          <span>
            Should Not Match Your Email Address
          </span>
        </li>
        <li>
          <span>
            1 Uppercase Character
          </span>
        </li>
        <li>
          <span>
            1 Number
          </span>
        </li>
      </ul>
    </div>
  </main>
</div>
`;
